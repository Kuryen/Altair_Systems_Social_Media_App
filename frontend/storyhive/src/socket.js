import { io } from "socket.io-client";

const socket = io("http://localhost:10000", {
  autoConnect: false,
}); // Replace with your actual server URL
export default socket;
