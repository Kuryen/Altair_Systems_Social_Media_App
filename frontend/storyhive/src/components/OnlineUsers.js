import React from "react";
import socket from "../socket"; // Ensure socket is imported here

function OnlineUsers({ onlineUsers, onSelectUser }) {
  const handleSelectUser = (user) => {
    const profileUsername = localStorage.getItem("profileUsername");
    // Check if profileUsername is not null or undefined
    if (profileUsername) {
      socket.emit("startChat", {
        from: profileUsername,
        to: user,
      });
      onSelectUser(user); // Notify parent component to set selected user
    }
  };

  return (
    <div className="w-1/4 bg-gray-100 p-4">
      <h2 className="text-xl font-bold mb-2">Online Users</h2>
      <ul>
        {onlineUsers.map((user, index) => (
          <li
            key={index}
            onClick={() => handleSelectUser(user)}
            className="cursor-pointer hover:bg-blue-200 p-2 rounded"
          >
            {user}
          </li>
        ))}
      </ul>
    </div>
  );
}

export default OnlineUsers;
