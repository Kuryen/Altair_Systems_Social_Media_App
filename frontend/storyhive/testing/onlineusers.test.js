import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import OnlineUsers from "../src/components/OnlineUsers";
import socket from "../src/socket";

// Mock socket.emit function
jest.mock("../src/socket", () => ({
  emit: jest.fn(),
}));

describe("OnlineUsers Component", () => {
  const onlineUsers = ["User1", "User2", "User3"];
  const onSelectUser = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    localStorage.setItem("profileUsername", "TestUser"); // Set profileUsername for each test
  });

  test("renders the 'Online Users' heading", () => {
    render(<OnlineUsers onlineUsers={[]} onSelectUser={onSelectUser} />);
    expect(screen.getByText("Online Users")).toBeInTheDocument();
  });

  test("renders the list of online users", () => {
    render(
      <OnlineUsers onlineUsers={onlineUsers} onSelectUser={onSelectUser} />
    );

    onlineUsers.forEach((user) => {
      expect(screen.getByText(user)).toBeInTheDocument();
    });
  });

  test("calls handleSelectUser and socket.emit with the correct arguments when a user is clicked", () => {
    render(
      <OnlineUsers onlineUsers={onlineUsers} onSelectUser={onSelectUser} />
    );

    const userElement = screen.getByText("User1");
    fireEvent.click(userElement);

    expect(onSelectUser).toHaveBeenCalledWith("User1");
    expect(socket.emit).toHaveBeenCalledTimes(1);
    expect(socket.emit).toHaveBeenCalledWith("startChat", {
      from: "TestUser", // Explicitly set expected profileUsername value
      to: "User1",
    });
  });

  test("does not throw an error when the onlineUsers list is empty", () => {
    render(<OnlineUsers onlineUsers={[]} onSelectUser={onSelectUser} />);
    expect(screen.getByText("Online Users")).toBeInTheDocument();
  });

  test("does not call socket.emit if localStorage profileUsername is missing", () => {
    localStorage.removeItem("profileUsername"); // Remove profileUsername

    render(
      <OnlineUsers onlineUsers={onlineUsers} onSelectUser={onSelectUser} />
    );

    const userElement = screen.getByText("User1");
    fireEvent.click(userElement);

    expect(socket.emit).not.toHaveBeenCalled();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <OnlineUsers onlineUsers={onlineUsers} onSelectUser={onSelectUser} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
