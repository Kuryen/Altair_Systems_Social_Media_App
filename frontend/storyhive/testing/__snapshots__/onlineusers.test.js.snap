// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OnlineUsers Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="onlineUsersContainer"
  >
    <h2>
      Online Users
    </h2>
    <ul>
      <li
        class="onlineUser"
      >
        User1
      </li>
      <li
        class="onlineUser"
      >
        User2
      </li>
      <li
        class="onlineUser"
      >
        User3
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
